### ISSUE TITLE
GetTokenInfo fetched twice in parallel

### ISSUE THREAD

**Issue by fleupold** [gnosis/dex-services#1155]
Looking at some kibana logs: https://logs.gnosisdev.com/goto/81b52cdb0cd9c7630dad320bdb4d525a

<img width="1287" alt="Screen Shot 2020-07-20 at 16 28 57" src="https://user-images.githubusercontent.com/1200333/87949300-260d2900-caa6-11ea-8e67-1ec9cfda8737.png">

We are sending the exact same RPC call twice with different request IDs. The reason might be similar as in #1154 that we check some cache, realize we need to go to network but do not block any parallel requests to also going to network.

This might also be causing the timeouts we see as we are putting a lot of pressure on the nodes (#1153)

**Comment by fleupold** [comment]
I took a look at this, and tried replacing the RwLock with a mutex that we hold on to throughout the cahce read, read through and write. However this would have the side effect, that we cannot read token info for different token infos in parallel (as they would be blocked on acquiring the lock).

I'm not sure this is worth it, and maybe we should just focus on batched reads rather than optimizing this race condition. Thoughts?

**Comment by e00E** [comment]
Having some requests twice is minor because it only happens the first time the token info is requested. We could fix this by calling TokenInfoCache::cache_all when we create the TokenInfoCache on startup before giving it to the price sources.

**Comment by nlordell** [comment]
I think we could also just check the cache again once we upgrade to a write lock.

**Comment by e00E** [comment]
That makes even more sense!

**Comment by e00E** [comment]
Based on comments in the above PR we can close this. 

> I'd prefer to not block the critical startup path for non-critical things such as token lookup for price estimates. Maybe querying twice is not so bad after all?

> The only way around this I see is to have signals (a mutex or oneshot channel) for each token being updated, that way each token is individually synchronized and we only need to hold the write lock when writing the update.

Imo the extra code isn't worth it for this.

**Comment by nlordell** [comment]
Closing as a "won't fix" because the impact is very small and not worth the added complexity.
